src/base/format_impl.h

```diff
@@ -0,0 +1,1016 @@
+				*out = '0';
+				++out;
+				++written;
+			} else {
+				size_t nr_digits = 0;
+				for (Number i = (arg < 0 ? -arg : arg); i; ++nr_digits, i /= base_) {
+				}
+
+				Number i = (arg < 0 ? -arg : arg);
+				for (size_t d = nr_digits; d; --d, i /= base_) {
+					static Number digit;
+					digit = (i % base_);
+					if (hexadecimal_ && digit > 9) {
+						*(out + d - 1) = (uppercase_ ? 'A' : 'a') + digit - 10;
+					} else {
+						*(out + d - 1) = '0' + digit;
+					}
+				}
+
+				out += nr_digits;
+				written += nr_digits;
+			}
+
+			if (written < min_width_) {
+				written = min_width_ - written;
+				for (; written; ++out, --written) {
+					*out = ' ';
+				}
+			}
+
+			return out;
+		}
+
+		// The more complex case: We want a right-aligned string with a given minimum width,
+		// padded with leading whitespace or zeroes.
+		size_t nr_digits;
+		if (arg == 0) {
+			nr_digits = 1;
+		} else {
+			nr_digits = 0;
+			for (Number i = (arg < 0 ? -arg : arg); i; ++nr_digits, i /= base_) {
+			}
+		}
+
+		size_t required_width = nr_digits;
+		if (arg < 0 || flags_ & kNumberSign) {
+			++required_width;
+		}
+		if (required_width < min_width_) {
+			required_width = min_width_ - required_width;
+			for (; required_width; ++out, --required_width) {
+				*out = (flags_ & kPadWith0 ? '0' : ' ');
+			}
+		}
+
+		if (arg < 0) {
+			*out = '-';
+			++out;
+		} else if (flags_ & kNumberSign) {
+			*out = '+';
+			++out;
+		}
+
+		if (arg == 0) {
+			*out = '0';
+			++out;
+		} else {
+			Number i = (arg < 0 ? -arg : arg);
+			for (size_t d = nr_digits; d; --d, i /= base_) {
+				static Number digit;
+				digit = (i % base_);
+				if (hexadecimal_ && digit > 9) {
+					*(out + d - 1) = (uppercase_ ? 'A' : 'a') + digit - 10;
+				} else {
+					*(out + d - 1) = '0' + digit;
+				}
+			}
+
+			out += nr_digits;
+		}
+
+		return out;
+	}
+
+	static const NumberNodeT node_;
+
+private:
+	const Number base_;
+	const bool hexadecimal_, uppercase_;
+};
+using IntNode = NumberNodeT<int64_t>;
+using UintNode = NumberNodeT<uint64_t>;
+template <typename NumberT>
+const NumberNodeT<NumberT> NumberNodeT<NumberT>::node_(kNone, 0, false, false);
+
+struct FloatNode : FormatNode {
+	FloatNode(const uint8_t f, const size_t w, const int32_t p) : FormatNode(f, w, p) {
+	}
+
+	char* append(char* out, const ArgType t, Argument arg_u, const bool localize) const override {
+		double arg;
+		switch (t) {
+		case ArgType::kFloat:
+			arg = arg_u.float_val;
+			break;
+		case ArgType::kSigned:
+			arg = arg_u.signed_val;
+			break;
+		case ArgType::kUnsigned:
+		case ArgType::kPointer:
+			arg = arg_u.unsigned_val;
+			break;
+		case ArgType::kNullptr:
+			arg_u.string_val = "nullptr";
+			return StringNode::node_.append(out, AbstractNode::ArgType::kString, arg_u, localize);
+		default:
+			throw wexception(
+			   "Wrong argument type: excepted float/double, found %s", to_string(t).c_str());
+		}
+
+		const int64_t as_int = static_cast<int64_t>(arg < 0 ? -arg : arg);
```

Comment:
Do I get it right that you don't implement `%f` for numbers with absolute value bigger than 2^63?
I understand that it's probably not needed in Widelands, but it should be documented and checked for.

Reply:
That's right; I added a comment

Reply:
Does no explicit check result in a standard exception for out of bounds conversion?
Should be tried with a test case?